src/Orleans/Messaging/ByteArrayBuilder.cs

```diff
@@ -148,8 +148,15 @@ private void EnsureRoomFor(int n)
         /// <returns></returns>
         public ByteArrayBuilder Append(byte[] array)
         {
+            int arrLen = array.Length;
             // Big enough for its own buffer
-            if (((currentOffset > MINIMUM_BUFFER_SIZE) && (array.Length > MINIMUM_BUFFER_SIZE)) || (array.Length >= bufferSize))
+            //
+            // This is a somewhat debatable optimization:
+            // 1) If the passed array is bigger than bufferSize, don't copy it and append it as its own buffer. 
+            // 2) Make sure to ALWAYS copy arrays which size is EXACTLY bufferSize, otherwise if the data was passed as an Immutable arg, 
+            // we may return this buffer back to the BufferPool and later over-write it.
+            // 3) If we already have MINIMUM_BUFFER_SIZE in the current buffer and passed enough data, also skip the copy and append it as its own buffer. 
+            if (((arrLen != bufferSize) && (currentOffset > MINIMUM_BUFFER_SIZE) && (arrLen > MINIMUM_BUFFER_SIZE)) || (arrLen > bufferSize))
```

Comment:
That one sign difference: arrLen > bufferSize instead of arrLen  >= bufferSize was a potential nasty bug!
Kudos to @jason-bragg for spotting it!